packages/itwin/tree-widget/README.md

```diff
@@ -218,47 +218,135 @@ function CustomModelsTreeComponent({ imodel, viewport, getSchemaContext, selecti
-  );
-}
-```
+  ```tsx
+  type UseModelsTreeProps = Parameters<typeof useModelsTree>[0];
+  type GetFilteredPathsType = Exclude<UseModelsTreeProps["getFilteredPaths"], undefined>;
+  interface CustomModelsTreeProps {
+    viewport: Viewport;
+    selectionStorage: SelectionStorage;
+    imodel: IModelConnection;
+    targetItems: InstanceKey[];
+  }
 
-<!-- END EXTRACTION -->
+  function CustomModelsTreeComponent({ viewport, selectionStorage, imodel, targetItems }: CustomModelsTreeProps) {
+    const getFilteredPaths = useCallback<GetFilteredPathsType>(
+      async ({ createInstanceKeyPaths }) => {
+        return createInstanceKeyPaths({
+          // list of instance keys representing nodes that should be displayed in the hierarchy
+          targetItems,
+        });
+      },
+      [targetItems],
+    );
+
+    const { modelsTreeProps, rendererProps } = useModelsTree({ activeView: viewport, getFilteredPaths });
+
+    return (
+      <VisibilityTree
+        {...modelsTreeProps}
+        getSchemaContext={getSchemaContext}
+        selectionStorage={selectionStorage}
+        imodel={imodel}
+        treeRenderer={(props) => <VisibilityTreeRenderer {...props} {...rendererProps} />}
+      />
+    );
+  }
+  ```
+
+  <!-- END EXTRACTION -->
+
+- **Post-process the paths created `createInstanceKeyPaths`**: Use `filter` string to generate the paths, then apply additional filtering - e.g., remove paths that are too long.
+  <!-- [[include: [TreeWidget.GetFilteredPathsComponentExample2], tsx]] -->
+  <!-- BEGIN EXTRACTION -->
+
+  ```tsx
+  interface CustomModelsTreeProps2 {
+    viewport: Viewport;
+    selectionStorage: SelectionStorage;
+    imodel: IModelConnection;
+  }
+
+  function CustomModelsTreeComponent2({ viewport, selectionStorage, imodel }: CustomModelsTreeProps2) {
+    const getFilteredPaths = useCallback<GetFilteredPathsType>(async ({ createInstanceKeyPaths, filter }) => {
+      const defaultPaths = await createInstanceKeyPaths({ label: filter ?? "test" });
+      const result = new Array<HierarchyFilteringPath>();
+      for (const path of defaultPaths) {
+        const normalizedPath = HierarchyFilteringPath.normalize(path);
+        if (normalizedPath.path.length < 5) {
+          normalizedPath.options = { autoExpand: true };
+          result.push(normalizedPath);
+        }
+      }
+      return result;
+    }, []);
+
+    const { modelsTreeProps, rendererProps } = useModelsTree({ activeView: viewport, getFilteredPaths });
+
+    return (
+      <VisibilityTree
+        {...modelsTreeProps}
+        getSchemaContext={getSchemaContext}
+        selectionStorage={selectionStorage}
+        imodel={imodel}
+        treeRenderer={(props) => <VisibilityTreeRenderer {...props} {...rendererProps} />}
+      />
+    );
+  }
+  ```
+
+  <!-- END EXTRACTION -->
+
+- **Apply fully custom logic**: Generate paths based on completely custom implementation. For exampel: showing each category that has an override applied in viewport.
```

Comment:
I don't think this is a good example - consumers have no way of knowing how we construct the hierarchy to create the paths by themselves. Instead, I think it would be better to demonstrate how to combine filtering by instance keys and by label (run a query to filter instance keys by given filter string and then pass them to `createInstanceKeyPaths`).